### PR TITLE
Alerting: Support `utf8_strict_mode: false` in Mimir

### LINKED ISSUE #87387: Alerting: Invalid syntax when creating Notification Policy on onprem Mimir Alertmanager
### What happened?

Whenever I try to create a Notification Policy I get following error: `"error validating Alertmanager config: bad matcher format: \"label_x\"=\"value_x\"\n"`

This happens with any combination of label/value

### What did you expect to happen?

A Notification Policy should be created

### Did this work before?

Yes, I believe it was the [UTF-8 support](https://github.com/grafana/grafana/pull/81512) that broke it, but I'm not sure

### How do we reproduce it?

1. Install Mimir with Alertmanager API enabled (tested 2.11 and 2.12)
2. Add the Alertmanager as Datasource
3. Create a Notification Policy with any label/value

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: Tested on both `Grafana Cloud v11.1.0-69372 (51da96d94e)` and OSS image id `c8b91775d855`
OS: Grafana Cloud and Docker
Browser: Chrome 123

### Grafana platform?

I use Grafana Cloud

### Datasource(s)?

Alertmanager

### LINKED ISSUE #89960: Alerting: Grafana UI cannot save notification policies in old versions of Mimir due to double quoting
There is a bug in the Grafana UI that prevents users for saving notification policies in old versions of Mimir, or current versions of Mimir with the `utf8-strict-mode` flag disabled.

This happens because we made change to the frontend where in recent versions of Grafana we double quote both the left and right hand side of matchers when saving notification policies. That is – we double quote both the label name and label value. For example: `"foo"="bar"`.

The problem is that users who are running Grafana 11 or newer with older versions of Mimir, or current versions of Mimir with the `utf8-strict-mode` flag disabled, will be unable to save their notification policies via the UI, because double quoting is not permitted on the left hand side of a matcher (label name) in older versions of Mimir or when `utf8-strict-mode` is disabled.

My recommendation on how to fix this is to only double quote on save if the text contains whitespace or any of the [reserved characters](https://github.com/prometheus/alertmanager/blob/main/matcher/parse/lexer.go#L27-L29), which are:

```
{ } ! = ~ , \ " ' `
```

The regular expression for that would look something like this:

```
[\s{}!=~,\\"'`]+
```

Another option is if a label name or label value doesn't match this regex then double quote it:

```
^[a-zA-Z_:][a-zA-Z0-9_:]*$
```

Where does this regex come from? It is the regex for [valid Prometheus label names](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels) (before UTF-8 support was added to Prometheus).

Either regex should work (although the first regex is what we use in Mimir, the second regex is more of an approximation), and will allow the UI to work with both old and new versions of Mimir, and also versions of Mimir that have `utf8-strict-mode` disabled. You do not need to change how the UI unquotes or parses matchers, and matchers should ALWAYS be unquoted (if quoted) when reading from the API.

Here are just a quick few examples of when double quoting is required:

```
foo="bar baz qux corge" // contains space
foo="bar, baz!" // contains both a comma and !
"foo bar"=baz // contains whitespace
"foo!"=baz // contains !
"foo\""=baz // contains "
"foo'"=baz // contains '
"foo\\"=baz // contains a \
"{foo}=baz // contains a { and a }
```

And here are some examples of where double quoting is not required:

```
foo=bar
_foo=bar
foo_=baz
:foo=bar
foo:=bar
foo123=baz456
```

You can see how the `^[a-zA-Z_:][a-zA-Z0-9_:]*$` regex works as an approximation.

**A simple rule**: Double quoting **IS ALWAYS ALLOWED** on the right-hand side of a matcher (the label value or regex). Double quoting **IS ONLY ALLOWED** on the left-hand side of a matcher when Alertmanager supports UTF-8. If it does not, or it is disabled, then the left-hand side of a matcher **MUST** match the regular expression `^[a-zA-Z_:][a-zA-Z0-9_:]*$`.

Related issues: #87387

### PR DESCRIPTION
**Pull request by gillesdemey**
**Why do we need this feature?**

This adds support for users who are running Mimir on-prem but have not yet opted in to `utf8_strict_mode`.

See https://github.com/grafana/grafana/issues/89960 for more context.

Fixes https://github.com/grafana/grafana/issues/89960 
Fixes https://github.com/grafana/grafana/issues/87387

### CODE REVIEW

**Comment by grobinson-grafana** [comment]
I tested this in dev, and it works for all expectations I tested ✅ 

The following table has three columns. The first two columns show the label name and pattern/label value I entered in the UI, and the last column shows the YAML that is stored in Mimir (read using mimirtool which just reads the raw YAML file).

Screenshot 2024-07-04 at 21 41 50

I did find a couple bugs in how the UI shows parses of these matchers when reading them from the API, but I'll open a separate issue for that as it's unrelated to this fix.

Nice work! 👍

**Review by grobinson-grafana — COMMENTED** [review]
LGTM! 👍 Please get a second review from @grafana/alerting-frontend for the code changes.

**Review by grobinson-grafana — APPROVED** [review]
LGTM! 👍 Please get a second review from @grafana/alerting-frontend for the code changes.

File: public/app/features/alerting/unified/utils/matchers.ts [hunk 1]
```diff
@@ -144,6 +145,22 @@ export function quoteWithEscape(input: string) {
   return `"${escaped}"`;
 }
 
+/**
+ * Quotes string only when reserved characters are used
+ */
+const RESERVED_CHARACTERS = /[\{\}\!\=\~\,\\\"\'\`\s]+/;
```
Comment (tomratcliffe) [inline public/app/features/alerting/unified/utils/matchers.ts:151]: What do you think about this? (It was fiddly to get exactly the same but it makes the characters potentially a little easier to read?)

```ts
const RESERVED_CHARACTERS = ['{', '}', '!', '=', '~', ',', '\\', '"', "'", '`', /\s/];
const RESERVED_CHARACTERS_REGEX = new RegExp(`[\\${RESERVED_CHARACTERS.join('\\')}]+`);
```

Comment (grobinson-grafana) [reply public/app/features/alerting/unified/utils/matchers.ts:151]: I'll let Gilles answer too, but for me I'm not sure it's worthwhile? Here are my thoughts:

1. The reserved characters are well documented in Alertmanager and will probably never change (without having to do another migration) as it's a breaking change to the grammar. I would be very surprised if you make changes to this regex in future.
2. I did ~2 hours of manual testing last night in dev using the existing regex and I'm confident in it 😄 If possible, I'd like to not have to do it again unless there is a really good reason for it.

Comment (tomratcliffe) [reply public/app/features/alerting/unified/utils/matchers.ts:151]: Sure, valid points :+1: I agree that it's unlikely we'd change the regex, but if it's nicer to read then its clearer what we're trying to do here.

This generated regex is the same as the one from before - it hasn't changed and the (unit) tests all pass with the different approach (with a small change to just make the formatting happier with the `\s`):

```ts
const RESERVED_CHARACTERS = ['{', '}', '!', '=', '~', ',', '\\', '"', "'", '`'];
const RESERVED_CHARACTERS_REGEX = new RegExp(`[\\${RESERVED_CHARACTERS.join('\\')}\\s]+`);
const original = /[\{\}\!\=\~\,\\\"\'\`\s]+/;

console.log(RESERVED_CHARACTERS_REGEX.toString() === original.toString()) // true
```

However - I won't die on this hill at all 😁

Comment (grobinson-grafana) [reply public/app/features/alerting/unified/utils/matchers.ts:151]: OK! This makes me feel better 😄 Gilles, up to you 👈

Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/matchers.ts:151]: I'll keep the regular expression, but I'll add a code comment that will help with deconstructing the array of special characters we have to pay attention to 👍

File: public/app/features/alerting/unified/utils/matchers.ts [hunk 2]
```diff
@@ -148,7 +148,11 @@ export function quoteWithEscape(input: string) {
 /**
  * Quotes string only when reserved characters are used
  */
+
+// The list of reserved characters that indicate we should be escaping the label key / value are
+// { } ! = ~ , \ " ' ` and any whitespace (\s), encoded in the regular expression below
```
Comment (grobinson-grafana) [inline public/app/features/alerting/unified/utils/matchers.ts:153]: You can link to this PR which has the grammar written out https://github.com/prometheus/alertmanager/pull/3453.

Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/matchers.ts:153]: Good idea! Let me append that

**Comment by grobinson-grafana** [comment]
Awesome! Let's 🚢 ?

**Review by tomratcliffe — APPROVED** [review]
LGTM! Thanks 👍

**Pull request merged**